packages/transactions/src/types.ts

```diff
@@ -4,23 +4,27 @@ import { IAccountMeta, IInstruction } from '@solana/instructions';
 export type SerializedMessageBytes = Uint8Array & { readonly __serializedMessageBytes: unique symbol };
 export type SerializedMessageBytesBase64 = string & { readonly __serializedMessageBytesBase64: unique symbol };
 
-export type BaseTransaction = Readonly<{
-    instructions: readonly IInstruction[];
-    version: TransactionVersion;
+export type BaseTransaction<
+    TVersion extends TransactionVersion = TransactionVersion,
+    TInstruction extends IInstruction = IInstruction
+> = Readonly<{
+    instructions: readonly TInstruction[];
+    version: TVersion;
 }>;
 
-type LegacyTransaction = BaseTransaction &
-    Readonly<{
-        instructions: readonly (Omit<IInstruction, 'accounts'> &
-            Readonly<{ readonly accounts?: readonly IAccountMeta[] }>)[];
-        version: 'legacy';
-    }>;
+type ILegacyInstruction<TProgramAddress extends string = string> = IInstruction<
+    TProgramAddress,
+    readonly IAccountMeta[]
+>;
+type LegacyTransaction<TInstruction extends ILegacyInstruction = ILegacyInstruction> = BaseTransaction<
+    'legacy',
+    TInstruction
+>;
 
-type V0Transaction = BaseTransaction &
-    Readonly<{
-        instructions: readonly IInstruction[];
-        version: 0;
-    }>;
+type V0Transaction<TInstruction extends IInstruction = IInstruction> = BaseTransaction<0, TInstruction>;
+
+export type Transaction<TInstruction extends IInstruction = IInstruction> =
+    | LegacyTransaction<TInstruction>
```

Comment:
I worry about how open this is. Is it, in fact, true that you can supply any arbitrary superset of `ILegacyInstruction` here, and expect it to parse as a legacy transaction's instruction?

Would it be better or worse to bake this `IAccountSignerMeta` type into this package, and enable/disable it with a ‘flag.’ Maybe that's bad for all the reason that configuration is usually bad? ([playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwgOwM7wGIEEByBhATQH0BlASQHEcBRAJVIAkyAVI6nLAIQBlqAROAF44JAJ4BbAEYQANgAoAlAG4AUCoCWyGMCgAzAIYJgcbvvHAsCJAFctcAN4q4zuPoAmbqMFSoAXHHQoTQBzJxcoWWB-AHJUdWDkHSIvdxQZUWi4AB84WPjEqCIAdyDtfUkZYEyc6JS3NIzs3JL1MoqqlQBfNU1tPUNjEmswHTwIWRJ8nWpkcsq3Sxs7Rxc4CMqYuISkuobq3K2C4tLgOY7Vw51-ewCRhHV9GU2Ya11dTO7ulRhREbhmKD6NCGGDqFAAHmYGCBCFEk22UBIAAtWnBQNpkG5UHA5D8RhBdHBsPhiOQqHRGCw2BwePwFE1bG5gLpNMA3EI4Izmay3AA+DkrFyaQLWBCglB+BxhVbOQxLGCSqEwuFTREo+Do4CY7F44AEom4QikSg0egkJisdhcXgCAD8ImGo3GMnhBRmZwWVggthgAG0ALpwfymcyLb1aAPSuCfNRyZDWGQyVzY2wAa2QECKyGT-0BwLFYOQCijADphTAoKLxWhS3LwwrVAB6RsuAB6tpUcYTSf0KeQ6cz2d7uaBqBBhfBuv1xKNZNNlMtNJtvOLqzLaArVcLqFrXp9qCbLec7aAA))

Reply:
My main reason for not wanting to bake that into `@solana/instructions` is too avoid a circular type dependency between the following 3 packages since:
- `@solana/signers` needs the `CompilableTransaction` type from the `@solana/transactions` package.
- `@solana/transactions` needs instruction types from `@solana/instructions`.

I also like the idea that no other packages in this library aside from `@solana/signers` are aware that signers exist. So it's this abstraction layer that simply sits on top of our current solution.

> I worry about how open this is. Is it, in fact, true that you can supply any arbitrary superset of ILegacyInstruction here, and expect it to parse as a legacy transaction's instruction?

I see what you mean. Technically this work because `readonly (IAccountMeta | IAccountLookupMeta)[]` is a superset of `readonly IAccountMeta[]` but we could remove the type parameters at this level to make it more explicit and less open like so:

```ts
type ILegacyInstruction<TProgramAddress extends string = string> = IInstruction<
    TProgramAddress,
    readonly IAccountMeta[]
>;
type LegacyTransaction = BaseTransaction<'legacy', ILegacyInstruction>;
type V0Transaction = BaseTransaction<0, IInstruction>;

export type Transaction = LegacyTransaction | V0Transaction;
```

This will still allow me to configure the `CompilableTransaction` type since it relied on `BaseTransaction` which is more open anyway.

What do you think?